twisted/protocols/tls.py

```diff
@@ -321,20 +315,32 @@ def makeConnection(self, transport):
         # Now that we ourselves have a transport (initialized by the
         # ProtocolWrapper.makeConnection call above), kick off the TLS
         # handshake.
+        self._checkHandshakeStatus()
+
+
+    def _checkHandshakeStatus(self):
+        """
+        Ask OpenSSL to proceed with a handshake in progress.
 
+        Initially, this just sends the ClientHello; after some bytes have been
+        stuffed in to the C{Connection} object by C{dataReceived}, it will then
+        respond to any C{Certificate} or C{KeyExchange} messages.
+        """
         # The connection might already be aborted (eg. by a callback during
         # connection setup), so don't even bother trying to handshake in that
         # case.
-        if not self._aborted:
-            try:
-                self._tlsConnection.do_handshake()
-            except WantReadError:
-                # This is the expected case - there's no data in the
-                # connection's input buffer yet, so it won't be able to
-                # complete the whole handshake now. If this is the speak-first
-                # side of the connection, then some bytes will be in the send
-                # buffer now; flush them.
-                self._flushSendBIO()
+        if self._aborted:
+            return
+        try:
+            self._tlsConnection.do_handshake()
+        except WantReadError:
+            self._flushSendBIO()
```

Comment:
I should note that we don't handle `WantWriteError` _anywhere_ in this file, and as far as I know, we never have.  As I understand it, `WantWriteError` is just OpenSSL's way of saying "`send()` set errno to `EAGAIN`", and since we're using memory BIOs here, OpenSSL isn't allowed to call `send()`.